File: client/src/crypto/CryptoTests.ts
import * as assert from 'assert';
import { NodeCrypto } from './NodeCrypto';
import { SECP256K1_FIELD_PRIME, SECP256K1_CURVE_ORDER } from './PointMath';

export interface TestReportItem {
    name: string;
    input: string;
    expected: string;
    actual: string;
    passed: boolean;
}

export function runCryptoUnitTests(): { allPassed: boolean; reports: TestReportItem[] } {
    console.log('=============================================================');
    console.log(' Stage 4: Cryptographic Primitives Validation (Client/Node)');
    console.log('=============================================================\n');

    let allPassed = true;
    const reports: TestReportItem[] = [];

    const recordTest = (name: string, input: string, expected: string, actual: string, condition: boolean) => {
        reports.push({
            name,
            input,
            expected,
            actual,
            passed: condition
        });
        console.log(`  - [${condition ? 'PASS' : 'FAIL'}] ${name} -> ${actual}`);
        if (!condition) allPassed = false;
    };

    // A. Secure random 32-byte values
    {
        const r1 = NodeCrypto.generateRandom32();
        const r2 = NodeCrypto.generateRandom32();
        const cond = r1.length === 32 && r2.length === 32 && !r1.equals(r2);
        recordTest(
            'A. Random 32-byte Generation',
            'None',
            '32 non-zero cryptographically secure bytes',
            `Len: ${r1.length}, Sample: ${r1.subarray(0, 4).toString('hex')}...`,
            cond
        );
    }

    // B. SHA-256
    {
        // Vector 1: Empty string
        const d1 = NodeCrypto.sha256(Buffer.alloc(0)).toString('hex');
        const exp1 = 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855';
        recordTest('B. SHA-256 (Empty string)', '""', exp1, d1, d1 === exp1);

        // Vector 2: "abc"
        const d2 = NodeCrypto.sha256(Buffer.from('abc')).toString('hex');
        const exp2 = 'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad';
        recordTest('B. SHA-256 ("abc")', '"abc"', exp2, d2, d2 === exp2);
    }

    // C. secp256k1 generator multiplication
    let pubG: Buffer = Buffer.alloc(0);
    let pub2G: Buffer = Buffer.alloc(0);
    let pub3G: Buffer = Buffer.alloc(0);
    {
        // 1 * G
        const priv1 = Buffer.alloc(32); priv1[31] = 1;
        pubG = NodeCrypto.scalarMultiplyBase(priv1);
        const expG = '0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798';
        recordTest('C. Generator mult (priv=1 -> G)', 'scalar=1', expG, pubG.toString('hex'), pubG.toString('hex') === expG);

        // 2 * G
        const priv2 = Buffer.alloc(32); priv2[31] = 2;
        pub2G = NodeCrypto.scalarMultiplyBase(priv2);
        const exp2G = '02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5';
        recordTest('C. Generator mult (priv=2 -> 2G)', 'scalar=2', exp2G, pub2G.toString('hex'), pub2G.toString('hex') === exp2G);

        // 3 * G
        const priv3 = Buffer.alloc(32); priv3[31] = 3;
        pub3G = NodeCrypto.scalarMultiplyBase(priv3);
        const exp3G = '02f9308a019258c31049344f85f89d5229b531c845836f99b08601f113bce036f9';
        recordTest('C. Generator mult (priv=3 -> 3G)', 'scalar=3', exp3G, pub3G.toString('hex'), pub3G.toString('hex') === exp3G);
    }

    // D. ECDH shared-secret calculation (Alice priv = 2, Bob priv = 3 -> 6G)
    {
        const privAlice = Buffer.alloc(32); privAlice[31] = 2;
        const privBob = Buffer.alloc(32); privBob[31] = 3;

        const secretAlice = NodeCrypto.ecdh(privAlice, pub3G);
        const secretBob = NodeCrypto.ecdh(privBob, pub2G);

        const exp6G = 'fff97bd5755eeea420453a14355235d382f6472f8568a18b2f057a1460297556';
        const match = secretAlice.equals(secretBob) && secretAlice.toString('hex') === exp6G;
        recordTest(
            'D. ECDH shared secret (2 * 3G == 3 * 2G == X(6G))',
            'Alice priv=2, Bob priv=3',
            exp6G,
            secretAlice.toString('hex'),
            match
        );
    }

    // E. EC point encoding
    {
        const validG = NodeCrypto.isValidPoint(pubG);
        const valid2G = NodeCrypto.isValidPoint(pub2G);
        const prefixOk = pubG[0] === 0x02 || pubG[0] === 0x03;
        const cond = validG && valid2G && prefixOk && pubG.length === 33;
        recordTest(
            'E. EC point encoding (33 bytes SEC1 compressed)',
            'Base point G',
            'Valid 33-byte SEC1 (prefix 0x02/0x03 + 32-byte X)',
            `Prefix: 0x0${pubG[0]}, len: ${pubG.length}`,
            cond
        );
    }

    // F. EC point addition (2G + 3G = 5G)
    {
        const sumPoint = NodeCrypto.pointAdd(pub2G, pub3G);
        const priv5 = Buffer.alloc(32); priv5[31] = 5;
        const pub5G = NodeCrypto.scalarMultiplyBase(priv5);
        const exp5G = '022f8bde4d1a07209355b4a7250a5c5128e88b84bddc619ab7cba8d569b240efe4';

        const match5G = sumPoint.equals(pub5G) && sumPoint.toString('hex') === exp5G;
        recordTest(
            'F. Point addition (2G + 3G == 5G)',
            'p1 = 2G, p2 = 3G',
            exp5G,
            sumPoint.toString('hex'),
            match5G
        );
    }

    // G. Modular arithmetic modulo p
    {
        const pMinus1Val = SECP256K1_FIELD_PRIME - 1n;
        const pMinus1Buf = Buffer.from(pMinus1Val.toString(16).padStart(64, '0'), 'hex');
        const twoBuf = Buffer.alloc(32); twoBuf[31] = 2;
        const oneBuf = Buffer.alloc(32); oneBuf[31] = 1;

        // 1. Add: (p - 1) + 2 mod p == 1
        const addRes = NodeCrypto.fieldAdd(pMinus1Buf, twoBuf);
        const expAdd = '0000000000000000000000000000000000000000000000000000000000000001';
        recordTest('G. Field Add ((p - 1) + 2 mod p == 1)', 'a = p - 1, b = 2', expAdd, addRes.toString('hex'), addRes.toString('hex') === expAdd);

        // 2. Sub: 1 - 2 mod p == p - 1
        const subRes = NodeCrypto.fieldSub(oneBuf, twoBuf);
        const expSub = pMinus1Buf.toString('hex');
        recordTest('G. Field Sub (1 - 2 mod p == p - 1)', 'a = 1, b = 2', expSub, subRes.toString('hex'), subRes.toString('hex') === expSub);

        // 3. Mul: (p - 1) * (p - 1) mod p == 1
        const mulRes = NodeCrypto.fieldMul(pMinus1Buf, pMinus1Buf);
        recordTest('G. Field Mul ((p - 1) * (p - 1) mod p == 1)', 'a = p - 1, b = p - 1', expAdd, mulRes.toString('hex'), mulRes.toString('hex') === expAdd);
    }

    // H. Negative / Boundary tests
    {
        // 1. Scalar = 0 rejected
        const zeroScalar = Buffer.alloc(32);
        let rejZero = !NodeCrypto.isValidScalar(zeroScalar);
        try {
            NodeCrypto.scalarMultiplyBase(zeroScalar);
            rejZero = false;
        } catch {}
        recordTest('H. Reject scalar == 0', 'scalar = 0', 'Rejected (false/throws)', rejZero ? 'Rejected' : 'Accepted', rejZero);

        // 2. Scalar = n rejected (order of curve)
        const nHex = SECP256K1_CURVE_ORDER.toString(16).padStart(64, '0');
        const nScalar = Buffer.from(nHex, 'hex');
        let rejN = !NodeCrypto.isValidScalar(nScalar);
        try {
            NodeCrypto.scalarMultiplyBase(nScalar);
            rejN = false;
        } catch {}
        recordTest('H. Reject scalar == n', 'scalar = n', 'Rejected (false/throws)', rejN ? 'Rejected' : 'Accepted', rejN);

        // 3. Scalar = p - 1 rejected as private scalar (since p - 1 > n)
        const pMinus1Hex = (SECP256K1_FIELD_PRIME - 1n).toString(16).padStart(64, '0');
        const pMinus1Scalar = Buffer.from(pMinus1Hex, 'hex');
        let rejPMinus1 = !NodeCrypto.isValidScalar(pMinus1Scalar);
        try {
            NodeCrypto.scalarMultiplyBase(pMinus1Scalar);
            rejPMinus1 = false;
        } catch {}
        recordTest('H. Reject scalar == p - 1 (since p - 1 > n)', 'scalar = p - 1', 'Rejected (false/throws)', rejPMinus1 ? 'Rejected' : 'Accepted', rejPMinus1);

        // 4. Malformed point prefix: 0x01
        const badPrefixPt = Buffer.from(pubG);
        badPrefixPt[0] = 0x01;
        const rejPrefix = !NodeCrypto.isValidPoint(badPrefixPt);
        recordTest('H. Reject point with invalid prefix 0x01', 'point prefix = 0x01', 'Rejected (false)', rejPrefix ? 'Rejected' : 'Accepted', rejPrefix);

        // 5. Point not on curve: X = 5 has no square root mod p
        const nonCurvePt = Buffer.alloc(33);
        nonCurvePt[0] = 0x02;
        nonCurvePt[32] = 0x05;
        const rejNonCurve = !NodeCrypto.isValidPoint(nonCurvePt);
        recordTest('H. Reject point not on curve (X=5 is non-quadratic residue)', 'X = 5', 'Rejected (false)', rejNonCurve ? 'Rejected' : 'Accepted', rejNonCurve);

        // 6. Wrong length scalar: 31 bytes
        const shortScalar = Buffer.alloc(31);
        const rejShortScalar = !NodeCrypto.isValidScalar(shortScalar);
        recordTest('H. Reject wrong-length scalar (31 bytes)', 'scalar len = 31', 'Rejected (false)', rejShortScalar ? 'Rejected' : 'Accepted', rejShortScalar);

        // 7. Wrong length point: 32 bytes
        const shortPoint = Buffer.alloc(32);
        const rejShortPoint = !NodeCrypto.isValidPoint(shortPoint);
        recordTest('H. Reject wrong-length point (32 bytes)', 'point len = 32', 'Rejected (false)', rejShortPoint ? 'Rejected' : 'Accepted', rejShortPoint);
    }

    console.log('\n=============================================================');
    console.log(` Stage 4 TypeScript Test Result: ${allPassed ? 'ALL PASSED' : 'FAILED'}`);
    console.log('=============================================================\n');

    return { allPassed, reports };
}

File: server/src/crypto/CryptoTests.cpp
#include "crypto/CryptoTests.hpp"
#include "crypto/TrezorCrypto.hpp"

#include <iostream>
#include <iomanip>
#include <cassert>
#include <cstring>
#include <vector>

namespace cot::crypto {

bool run_crypto_unit_tests() {
    std::cout << "=========================================================\n";
    std::cout << " Stage 4: Cryptographic Primitives Validation (Server/C++)\n";
    std::cout << "=========================================================\n\n";

    bool all_passed = true;

    auto check_test = [&](const std::string& name, bool condition, const std::string& info = "") {
        std::cout << "  - [" << (condition ? "PASS" : "FAIL") << "] " << name;
        if (!info.empty()) {
            std::cout << " -> " << info;
        }
        std::cout << "\n";
        if (!condition) all_passed = false;
    };

    // A. Secure Random 32-byte Values
    {
        uint8_t r1[32] = {0};
        uint8_t r2[32] = {0};
        generate_random_32(r1);
        generate_random_32(r2);

        bool r1_nonzero = false;
        bool different = false;
        for (size_t i = 0; i < 32; ++i) {
            if (r1[i] != 0) r1_nonzero = true;
            if (r1[i] != r2[i]) different = true;
        }
        check_test("A. Random 32-byte Generation", r1_nonzero && different,
                   "Len: 32, Entropy: arc4random_buf via Trezor random_buffer hook");
    }

    // B. SHA-256
    {
        // NIST Vector 1: ""
        uint8_t d1[32];
        sha256(nullptr, 0, d1);
        std::string h1 = to_hex(d1, 32);
        check_test("B. SHA-256 (Empty string)", 
                   h1 == "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
                   h1);

        // NIST Vector 2: "abc"
        const uint8_t abc[] = {'a', 'b', 'c'};
        uint8_t d2[32];
        sha256(abc, 3, d2);
        std::string h2 = to_hex(d2, 32);
        check_test("B. SHA-256 ('abc')", 
                   h2 == "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad",
                   h2);
    }

    // C. secp256k1 generator multiplication
    uint8_t pub_g[33] = {0};
    uint8_t pub_2g[33] = {0};
    uint8_t pub_3g[33] = {0};
    {
        // Scalar 1 -> G
        uint8_t priv1[32] = {0};
        priv1[31] = 1;
        bool ok1 = scalar_multiply_base(priv1, pub_g);
        std::string h_g = to_hex(pub_g, 33);
        check_test("C. Generator mult (priv=1 -> G)",
                   ok1 && h_g == "0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798",
                   h_g);

        // Scalar 2 -> 2G
        uint8_t priv2[32] = {0};
        priv2[31] = 2;
        bool ok2 = scalar_multiply_base(priv2, pub_2g);
        std::string h_2g = to_hex(pub_2g, 33);
        check_test("C. Generator mult (priv=2 -> 2G)",
                   ok2 && h_2g == "02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5",
                   h_2g);

        // Scalar 3 -> 3G
        uint8_t priv3[32] = {0};
        priv3[31] = 3;
        bool ok3 = scalar_multiply_base(priv3, pub_3g);
        std::string h_3g = to_hex(pub_3g, 33);
        check_test("C. Generator mult (priv=3 -> 3G)",
                   ok3 && h_3g == "02f9308a019258c31049344f85f89d5229b531c845836f99b08601f113bce036f9",
                   h_3g);
    }

    // D. ECDH shared-secret calculation (Alice priv = 2, Bob priv = 3 -> 6G)
    {
        uint8_t priv_alice[32] = {0}; priv_alice[31] = 2;
        uint8_t priv_bob[32] = {0};   priv_bob[31] = 3;

        uint8_t secret_alice[32] = {0};
        uint8_t secret_bob[32] = {0};

        bool ok_a = ecdh(priv_alice, pub_3g, secret_alice);
        bool ok_b = ecdh(priv_bob, pub_2g, secret_bob);

        std::string h_sa = to_hex(secret_alice, 32);
        std::string h_sb = to_hex(secret_bob, 32);
        bool match = (ok_a && ok_b && h_sa == h_sb &&
                      h_sa == "fff97bd5755eeea420453a14355235d382f6472f8568a18b2f057a1460297556");
        check_test("D. ECDH shared secret (2 * 3G == 3 * 2G == X(6G))",
                   match,
                   "Normalized 32-byte X: " + h_sa);
    }

    // E. EC point encoding
    {
        bool valid_g = is_valid_point(pub_g);
        bool valid_2g = is_valid_point(pub_2g);
        bool prefix_ok = (pub_g[0] == 0x02 || pub_g[0] == 0x03);
        check_test("E. EC point encoding (33 bytes SEC1 compressed)",
                   valid_g && valid_2g && prefix_ok,
                   "Prefix: 0x0" + std::to_string(pub_g[0]) + ", len: 33");
    }

    // F. EC point addition (2G + 3G = 5G)
    {
        uint8_t sum_point[33] = {0};
        bool ok_add = point_add(pub_2g, pub_3g, sum_point);
        std::string h_sum = to_hex(sum_point, 33);

        // Expected 5G from scalar mult
        uint8_t priv5[32] = {0}; priv5[31] = 5;
        uint8_t pub_5g[33] = {0};
        scalar_multiply_base(priv5, pub_5g);
        std::string h_5g = to_hex(pub_5g, 33);

        bool match_5g = (ok_add && h_sum == h_5g &&
                         h_sum == "022f8bde4d1a07209355b4a7250a5c5128e88b84bddc619ab7cba8d569b240efe4");
        check_test("F. Point addition (2G + 3G == 5G)",
                   match_5g,
                   h_sum);
    }

    // G. Modular arithmetic modulo p
    {
        // Field prime p in 32 bytes
        const auto& p_bytes = SECP256K1_FIELD_PRIME_BE;

        // p - 1
        uint8_t p_minus_1[32];
        std::memcpy(p_minus_1, p_bytes.data(), 32);
        p_minus_1[31] = 0x2E; // 0x2F - 1 = 0x2E

        // Scalar 2
        uint8_t two[32] = {0};
        two[31] = 2;

        // 1. Add: (p - 1) + 2 mod p == 1
        uint8_t add_res[32] = {0};
        field_add(p_minus_1, two, add_res);
        std::string h_add = to_hex(add_res, 32);
        check_test("G. Field Add ((p - 1) + 2 mod p == 1)",
                   h_add == "0000000000000000000000000000000000000000000000000000000000000001",
                   h_add);

        // 2. Sub: 1 - 2 mod p == p - 1
        uint8_t one[32] = {0};
        one[31] = 1;
        uint8_t sub_res[32] = {0};
        field_sub(one, two, sub_res);
        std::string h_sub = to_hex(sub_res, 32);
        std::string h_p_minus_1 = to_hex(p_minus_1, 32);
        check_test("G. Field Sub (1 - 2 mod p == p - 1)",
                   h_sub == h_p_minus_1,
                   h_sub);

        // 3. Mul: (p - 1) * (p - 1) mod p == (-1)*(-1) == 1
        uint8_t mul_res[32] = {0};
        field_mul(p_minus_1, p_minus_1, mul_res);
        std::string h_mul = to_hex(mul_res, 32);
        check_test("G. Field Mul ((p - 1) * (p - 1) mod p == 1)",
                   h_mul == "0000000000000000000000000000000000000000000000000000000000000001",
                   h_mul);
    }

    // H. Negative / Boundary tests
    {
        // 1. Scalar = 0 rejected
        uint8_t zero_scalar[32] = {0};
        uint8_t dummy_out[33];
        bool rej_zero = !scalar_multiply_base(zero_scalar, dummy_out) && !is_valid_scalar(zero_scalar);
        check_test("H. Reject scalar == 0", rej_zero);

        // 2. Scalar = n rejected (order of curve)
        const auto& n_bytes = SECP256K1_CURVE_ORDER_BE;
        bool rej_n = !scalar_multiply_base(n_bytes.data(), dummy_out) && !is_valid_scalar(n_bytes.data());
        check_test("H. Reject scalar == n", rej_n);

        // 3. Scalar = p - 1 rejected as scalar (since p - 1 > n), but valid in field
        uint8_t p_minus_1[32];
        std::memcpy(p_minus_1, SECP256K1_FIELD_PRIME_BE.data(), 32);
        p_minus_1[31] = 0x2E;
        bool rej_p_minus_1 = !scalar_multiply_base(p_minus_1, dummy_out) && !is_valid_scalar(p_minus_1);
        check_test("H. Reject scalar == p - 1 (since p - 1 > n)", rej_p_minus_1);

        // 4. Malformed point prefix: 0x01
        uint8_t bad_prefix_pt[33];
        std::memcpy(bad_prefix_pt, pub_g, 33);
        bad_prefix_pt[0] = 0x01;
        bool rej_prefix = !is_valid_point(bad_prefix_pt);
        check_test("H. Reject point with invalid prefix 0x01", rej_prefix);

        // 5. Point not on curve: X = 5 has no square root mod p (non-quadratic residue)
        uint8_t non_curve_pt[33] = {0};
        non_curve_pt[0] = 0x02;
        non_curve_pt[32] = 0x05;
        bool rej_non_curve = !is_valid_point(non_curve_pt);
        check_test("H. Reject point not on curve (X=5 is non-quadratic residue)", rej_non_curve);
    }

    std::cout << "\n=========================================================\n";
    std::cout << " Stage 4 C++ Test Result: " << (all_passed ? "ALL PASSED" : "FAILED") << "\n";
    std::cout << "=========================================================\n\n";

    return all_passed;
}

} // namespace cot::crypto
